lib/grppcatr.gi

```diff
@@ -480,18 +480,31 @@ end );
 ##
 #M  MaximalNormalSubgroups( <G> )
 ##
-InstallMethod( MaximalNormalSubgroups, "for solvable groups",
+InstallMethod( MaximalNormalSubgroups, "for abelian groups",
                [ IsGroup and IsAbelian ],
+               # IsGroup and IsFinite ranks higher than
+               # IsGroup and IsAbelian, so we have to increase the
+               # rank, otherwise the method for normal subgroup computation
+               # is selected.
+               RankFilter( IsGroup and IsFinite and IsAbelian )
+               - RankFilter( IsGroup and IsAbelian ),
 function( G )
     local Gf,     # FactorGroup of G
           hom,    # homomorphism from G to Gf
-          MaxGf;  # MaximalNormalSubgroups of Gf
+          MaxGf,  # MaximalNormalSubgroups of Gf
+          AbInv;  # Abelian invariants of G
     if not IsPcGroup(G) then
-        # convert it to an Abelian PcGroup with same invariants
-        Gf := AbelianGroup(IsPcGroup, AbelianInvariants(G));
-        hom := IsomorphismGroups(G, Gf);
-        MaxGf := NormalMaximalSubgroups(Gf);
-        return List(MaxGf, N -> PreImage(hom, N));
+        AbInv := AbelianInvariants(G);
+        if 0 in AbInv then
+            # (p) is a maximal normal subgroup in Z for every prime p
+            return fail;
+        else
+            # convert it to an Abelian PcGroup with same invariants
+            Gf := AbelianGroup(IsPcGroup, AbInv);
+            hom := IsomorphismGroups(G, Gf);
```

Comment:
Computing `IsomorphismGroups` might be wasteful here... Why not use `IsomorphismPcGroup`? I.e. replace the preceding two lines by
```
hom := IsomorphismPcGroup(G);
Gf := Image(hom);